README.md
# HAMMING ENC/DEC

Familiarization with the Hamming method of error-correcting coding, which allows to detect and automatically correct errors that occur during the transmission of information.

# USAGE

```sh
./hammingenc input.file output.hmng [8/12/16/32/64]
./hammingdec input.hmng output.file [8/12/16/32/64]
```

2019


CHANGELOG
https://github.com/zagalskij/HammingCode

1.3: zvezdochiot <mykaralw@yandex.ru>

    * magic word
    * restruct for splint

1.2: zvezdochiot <mykaralw@yandex.ru>

    * added warnings
    * optimize io, time gain 50-70%

1.1: zvezdochiot <mykaralw@yandex.ru>

    * simplify invert
    * simplify hammingdec
    * strip code
    * bit operations, remove math, time gain 50-70%

1.0: Dmitrii Zagalskij

    * init version


union_heming.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void kontrolbit(int *kod, int lengh)
{
    int i = 0, summ = 0;
    while(i < lengh)
    {
        if((i+1 == 1)||(i+1 == 2)||(i+1 == 4)||(i+1 == 8)||(i+1 == 16)||(i+1 == 32)||(i+1 == 64))
        {
            int start = i;
            while(start < lengh)
            {
                for(int g = 0; g < i + 1; g++)
                {
                    if((kod[start+g] != 2)&&(start+g < lengh))
                        summ = summ + kod[start+g];
                }
                start = start + i + i + 2;
            }
            kod[i] = summ%2;
            summ = 0;
        }
        i++;
    }
}

void tobit(int x, int *byte)
{
    for(int i = 0; i < 8; i++)
    {
        byte[i] = x%2;
        x = x/2;
    }
    invert(byte);
}

void invert(int *byte)
{
    int tmp[8], g = 0;
    for(int i = 7; i > -1; i--)
        tmp[g++] = byte[i];
    for(int i = 0; i < 8; i++)
        byte[i] = tmp[i];
}

int todec(int *byte)
{
    int x = 0, st = 7;
    for(int i = 0; i < 8; i++)
    {
        x = x + byte[i]*pow(2, st);
        st--;
    }
    return x;
}

int main(int argc, char *argv[])
{
    if (argc < 3)
    {
        printf("Usage: %s input.file output.hmng [8/12/16/32/64]\n", argv[0]);
        return 0;
    }
    FILE *f = fopen(argv[1], "rb");
    FILE *t = fopen(argv[2], "w+b");
    int byte[8];
    printf("file: %s\n", argv[1]);
    printf("coded: %s\n", argv[2]);
    int c = 8;
    if (argc > 3) c = atoi(argv[3]);
    printf("coding length: %d\n\n", c);
    if(c == 8)
    {
        int x, kod[12];
        while(fgetc(f) != EOF)
        {
            fseek(f, -1, SEEK_CUR);
            x = fgetc(f);
            tobit(x, byte);
            int g = 0;
            for(int i = 0; i < 12; i++)
            {
                if((i != 0)&&(i != 1)&&(i != 3)&&(i != 7))
                    kod[i] = byte[g++];
                else
                    kod[i] = 2;
            }
            kontrolbit(kod, 12);
            g = 0;
            while(g < 12)
            {
                for(int i = 0; i < 8; i++)
                {
                    if((i == 0)||(i == 1))
                        byte[i] = 0;
                    else
                        byte[i] = kod[g++];
                }
                fputc(todec(byte), t);
            }
        }
    }
    else if(c == 12)
    {
        int x, kod[17], g = 0, flag = 0;
        while(fgetc(f) != EOF)
        {
            if(flag == 0)
            {
                fseek(f, -1, SEEK_CUR);
                x = fgetc(f);
                tobit(x, byte);
                for(int i = 0; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                x = fgetc(f);
                tobit(x, byte);
                for(int i = 0; i < 4; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                fseek(f, -1, SEEK_CUR);
                flag = 1;
            }
            else if(flag == 1)
            {
                fseek(f, -1, SEEK_CUR);
                x = fgetc(f);
                tobit(x, byte);
                for(int i = 4; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                x = fgetc(f);
                tobit(x, byte);
                for(int i = 0; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                flag = 0;
            }
            if(g == 17)
            {
                kontrolbit(kod, 17);
                g = 0;
                while(g < 17)
                {
                    for(int i = 0; i < 8; i++)
                    {
                        if(i != 7)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
        }
    }
    else if(c == 16)
    {
        int x, kod[21], g = 0;
        while(fgetc(f) != EOF)
        {
            fseek(f, -1, SEEK_CUR);
            x = fgetc(f);
            tobit(x, byte);
            for(int i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 21)
            {
                kontrolbit(kod, 21);
                g = 0;
                while(g < 21)
                {
                    for(int i = 0; i < 8; i++)
                    {
                        if(i == 0)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
        }
    }
    else if(c == 32)
    {
        int x, kod[38], g = 0;
        while(fgetc(f) != EOF)
        {
            fseek(f, -1, SEEK_CUR);
            x = fgetc(f);
            tobit(x, byte);
            for(int i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 38)
            {
                kontrolbit(kod, 38);
                g = 0;
                while(g < 38)
                {
                    for(int i = 0; i < 8; i++)
                    {
                        if((i != 6)&&(i != 7))
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
        }
    }
    else if(c == 64)
    {
        int x, kod[71], g = 0;
        while(fgetc(f) != EOF)
        {
            fseek(f, -1, SEEK_CUR);
            x = fgetc(f);
            tobit(x, byte);
            for(int i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 71)
            {
                kontrolbit(kod, 71);
                g = 0;
                while(g < 71)
                {
                    for(int i = 0; i < 8; i++)
                    {
                        if(i != 7)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
        }
    }
    else
    {
        printf("Error: bad coding length %d, maybe 8/12/16/32/64\n\n", c);
    }
    fclose(f);
    fclose(t);
    return 0;
}


src/hammingdec.c
#include <stdio.h>
#include <stdlib.h>

void kontrolbit(int *kod, int lengh)
{
    int i = 0, summ = 0, errorpos = 0, start, g;
    while(i < lengh)
    {
        if((i+1 == 1)||(i+1 == 2)||(i+1 == 4)||(i+1 == 8)||(i+1 == 16)||(i+1 == 32)||(i+1 == 64))
        {
            start = i;
            while(start < lengh)
            {
                for(g = 0; g < i + 1; g++)
                {
                    if((start+g != i)&&(start+g < lengh))
                        summ += kod[start+g];
                }
                start += (i + i + 2);
            }
            if(kod[i] != summ%2)
            {
                errorpos += i;
            }
            summ = 0;
        }
        i++;
    }
    kod[errorpos] = (errorpos != 0) ? !kod[errorpos] : kod[errorpos];
}

void invert(int *byte)
{
    int i, t;
    for(i = 0; i < 4; i++)
    {
        t = byte[i];
        byte[i] = byte[7 - i];
        byte[7 - i] = t;
    }
}

void tobit(int x, int *byte)
{
    int i;
    for(i = 0; i < 8; i++)
    {
        byte[i] = x%2;
        x >>= 1;
    }
    invert(byte);
}

int todec(int *byte)
{
    int i, x = 0, st = 7;
    for(i = 0; i < 8; i++)
    {
        x += (byte[i]<<st);
        st--;
    }
    return x;
}

int main(int argc, char *argv[])
{
    FILE *f, *t;
    int byte[8];
    int c = 8;
    char mc[7] = {"\0"};
    int x, i, kod[71], g;
    int tmp[4], flag, schet;
    if (argc < 3)
    {
        printf("Usage: %s input.hmng output.file [8/12/16/32/64]\n", argv[0]);
        return 0;
    }
    f = fopen(argv[1], "rb");
    t = fopen(argv[2], "w+b");
    printf("coded: %s\n", argv[1]);
    printf("file: %s\n", argv[2]);
    fread(mc, 6, 1, f);
    if (sscanf(mc, "HECC%d", &c) <= 0)
    {
        fprintf(stderr, "WARNING: Bad MAGIC word: %s\n\n", mc);
    }
    if (argc > 3) c = atoi(argv[3]);
    printf("decoding length: %d\n\n", c);
    if(c == 8)
    {
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if((i != 0)&&(i != 1))
                    kod[g++] = byte[i];
            }
            if(g == 12)
            {
                kontrolbit(kod, 12);
                g = 0;
                for(i = 0; i < 12; i++)
                {
                    if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8))
                    {
                        byte[g++] = kod[i];
                    }
                }
                fputc(todec(byte), t);
                g = 0;
            }
            x = fgetc(f);
        }
    }
    else if(c == 12)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        flag = 0;
        schet = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            kod[g] = byte[7];
            g++;
            if(g == 17)
            {
                g = 0;
                kontrolbit(kod, 17);
                g = 0;
                if(flag == 1) //костыль
                {
                    for(i = 0; i < 4; i++)
                    {
                        byte[g++] = tmp[i];
                    }
                    flag = 0;
                }
                for(i = 0; i < 17; i++)
                {
                    if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
                        byte[g++] = kod[i];
                    if(g == 8)
                    {
                        fputc(todec(byte), t);
                        g = 0;
                        if((flag == 0)&&(schet%2 == 0)) //костыль 2
                        {
                            for(i = 12; i < 17; i++)
                            {
                                if(i+1 != 16)
                                {
                                    tmp[g++] = kod[i];
                                }
                            }
                            flag = 1;
                        }
                        g = 0;
                    }
                }
                g = 0;
            }
            schet++; //костыль 3
            x = fgetc(f);
        }
    }
    else if(c == 16)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if(i != 0)
                    kod[g++] = byte[i];
            }
            if(g == 21)
            {
                kontrolbit(kod, 21);
                g = 0;
                for(i = 0; i < 21; i++)
                {
                    if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
                        byte[g++] = kod[i];
                    if(g == 8)
                    {
                        fputc(todec(byte), t);
                        g = 0;
                    }
                }
            }
            x = fgetc(f);
        }
    }
    else if(c == 32)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if((i == 6)||(i == 7))
                    kod[g++] = byte[i];
            }
            if(g == 38)
            {
                kontrolbit(kod, 38);
                g = 0;
                for(i = 0; i < 38; i++)
                {
                    if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
                        byte[g++] = kod[i];
                    if(g == 8)
                    {
                        fputc(todec(byte), t);
                        g = 0;
                    }
                }
            }
            x = fgetc(f);
        }
    }
    else if(c == 64)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if(i == 7)
                    kod[g++] = byte[i];
            }
            if(g == 71)
            {
                kontrolbit(kod, 71);
                g = 0;
                for(i = 0; i < 71; i++)
                {
                    if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
                        byte[g++] = kod[i];
                    if(g == 8)
                    {
                        fputc(todec(byte), t);
                        g = 0;
                    }
                }
            }
            x = fgetc(f);
        }
    }
    else
    {
        fprintf(stderr, "ERROR: bad coding length %d, maybe 8/12/16/32/64\n\n", c);
    }
    fclose(f);
    fclose(t);
    return 0;
}


union_deheming.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int todec(int *byte);
void invert(int *byte);
void tobit(int x, int *byte);
void kontrolbit(int *kod, int lengh);

int main()
{
	FILE *f = fopen("C:/laba/2.txt", "rb");
	FILE *t = fopen("C:/laba/3.txt", "w+b");
	int byte[8];
	printf("coded file: C:/laba/2.txt\n");
	printf("file: C:/laba/3.txt\n\n");
	printf("select decoding length 8/12/16/32/64: ");
	int c;
	scanf("%d", &c);
	if(c == 8)
	{
		int x, kod[12], g = 0;
		while(fgetc(f) != EOF)
		{
			fseek(f, -1, SEEK_CUR);
			x = fgetc(f);
			tobit(x, byte);
			for(int i = 0; i < 8; i++)
			{
				if((i != 0)&&(i != 1))
					kod[g++] = byte[i];
			}
			if(g == 12)
			{
				kontrolbit(kod, 12);
				g = 0;
				for(int i = 0; i < 12; i++)
				{
					if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8))
					{
						byte[g++] = kod[i];
					}
				}
				fputc(todec(byte), t);
				g = 0;
			}
		}
	}
	else if(c == 12)
	{
		int tmp[4], x, kod[17], g = 0, flag = 0, schet = 0;
		while(fgetc(f) != EOF)
		{
			fseek(f, -1, SEEK_CUR);
			x = fgetc(f);
			tobit(x, byte);
			kod[g] = byte[7];
			g++;
			if(g == 17)
			{
				g = 0;
				kontrolbit(kod, 17);
				g = 0;
				if(flag == 1) //костыль
				{
					for(int i = 0; i < 4; i++)
					{
						byte[g++] = tmp[i];
					}
					flag = 0;
				}
				for(int i = 0; i < 17; i++)
				{
					if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
						byte[g++] = kod[i];
					if(g == 8)
					{
						fputc(todec(byte), t);
						g = 0;
						if((flag == 0)&&(schet%2 == 0)) //костыль 2
						{
							for(int i = 12; i < 17; i++) 
							{
								if(i+1 != 16)
								{
									tmp[g++] = kod[i];
								}
							}
							flag = 1;
						}
						g = 0;
					}
				}
				g = 0;
			}
			schet++; //костыль 3
		}
	}
	else if(c == 16)
	{
		int x, kod[21], g = 0; 
		while(fgetc(f) != EOF)
		{
			fseek(f, -1, SEEK_CUR);
			x = fgetc(f);
			tobit(x, byte);
			for(int i = 0; i < 8; i++)
			{
				if(i != 0)
					kod[g++] = byte[i];
			}
			if(g == 21)
			{
				kontrolbit(kod, 21);
				g = 0;
				for(int i = 0; i < 21; i++)
				{
					if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
						byte[g++] = kod[i];
					if(g == 8)
					{
						fputc(todec(byte), t);
						g = 0;
					}
				}
			}
		}
	}
	else if(c == 32)
	{
		int x, kod[38], g = 0; 
		while(fgetc(f) != EOF)
		{
			fseek(f, -1, SEEK_CUR);
			x = fgetc(f);
			tobit(x, byte);
			for(int i = 0; i < 8; i++)
			{
				if((i == 6)||(i == 7))
					kod[g++] = byte[i];
			}
			if(g == 38)
			{
				kontrolbit(kod, 38);
				g = 0;
				for(int i = 0; i < 38; i++)
				{
					if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
						byte[g++] = kod[i];
					if(g == 8)
					{
						fputc(todec(byte), t);
						g = 0;
					}
				}
			}
		}
	}
	else if(c == 64)
	{
		int x, kod[71], g = 0; 
		while(fgetc(f) != EOF)
		{
			fseek(f, -1, SEEK_CUR);
			x = fgetc(f);
			tobit(x, byte);
			for(int i = 0; i < 8; i++)
			{
				if(i == 7)
					kod[g++] = byte[i];
			}
			if(g == 71)
			{
				kontrolbit(kod, 71);
				g = 0;
				for(int i = 0; i < 71; i++)
				{
					if((i+1 != 1)&&(i+1 != 2)&&(i+1 != 4)&&(i+1 != 8)&&(i+1 != 16)&&(i+1 != 32)&&(i+1 != 64))
						byte[g++] = kod[i];
					if(g == 8)
					{
						fputc(todec(byte), t);
						g = 0;
					}
				}
			}
		}
	}
	fclose(f);
	fclose(t);
	printf("COMPLETE!");
	getchar(); getchar();
	return 0;
}

void kontrolbit(int *kod, int lengh)
{
	int i = 0, summ = 0, errorpos = 0;
	while(i < lengh)
	{
		if((i+1 == 1)||(i+1 == 2)||(i+1 == 4)||(i+1 == 8)||(i+1 == 16)||(i+1 == 32)||(i+1 == 64))
		{
			int start = i;
			while(start < lengh)
			{
				for(int g = 0; g < i + 1; g++)
				{
					if((start+g != i)&&(start+g < lengh))
						summ = summ + kod[start+g];
				}
				start = start + i + i + 2;
			}
			if(kod[i] != summ%2)
			{
				errorpos = errorpos + i;
			}
			summ = 0;
		}
		i++;
	}
	if(errorpos != 0)
	{
		if(kod[errorpos] == 0)
			kod[errorpos] = 1;
		else
			kod[errorpos] = 0;
	}
}

void tobit(int x, int *byte)
{
	for(int i = 0; i < 8; i++)
	{
		byte[i] = x%2;
		x = x/2;
	}
	invert(byte);
}

void invert(int *byte)
{
	int tmp[8], g = 0;
	for(int i = 7; i > -1; i--)
		tmp[g++] = byte[i];
	for(int i = 0; i < 8; i++)
		byte[i] = tmp[i];
}

int todec(int *byte)
{
	int x = 0, st = 7;
	for(int i = 0; i < 8; i++)
	{
		x = x + byte[i]*pow(2, st);
		st--;
	}
	return x;
}

src/hammingenc.c
#include <stdio.h>
#include <stdlib.h>

void kontrolbit(int *kod, int lengh)
{
    int i = 0, summ = 0, start, g;
    while(i < lengh)
    {
        if((i+1 == 1)||(i+1 == 2)||(i+1 == 4)||(i+1 == 8)||(i+1 == 16)||(i+1 == 32)||(i+1 == 64))
        {
            start = i;
            while(start < lengh)
            {
                for(g = 0; g < i + 1; g++)
                {
                    if((kod[start+g] != 2)&&(start+g < lengh))
                        summ += kod[start+g];
                }
                start += (i + i + 2);
            }
            kod[i] = summ%2;
            summ = 0;
        }
        i++;
    }
}

void invert(int *byte)
{
    int i, t;
    for(i = 0; i < 4; i++)
    {
        t = byte[i];
        byte[i] = byte[7 - i];
        byte[7 - i] = t;
    }
}

void tobit(int x, int *byte)
{
    int i;
    for(i = 0; i < 8; i++)
    {
        byte[i] = x%2;
        x >>= 1;
    }
    invert(byte);
}

int todec(int *byte)
{
    int i, x = 0, st = 7;
    for(i = 0; i < 8; i++)
    {
        x += (byte[i]<<st);
        st--;
    }
    return x;
}

int main(int argc, char *argv[])
{
    FILE *f, *t;
    int byte[8];
    int c = 8;
    char mc[2];
    int x, kod[71], i, g, flag;
    if (argc < 3)
    {
        printf("Usage: %s input.file output.hmng [8/12/16/32/64]\n", argv[0]);
        return 0;
    }
    f = fopen(argv[1], "rb");
    t = fopen(argv[2], "w+b");
    printf("file: %s\n", argv[1]);
    printf("coded: %s\n", argv[2]);
    if (argc > 3) c = atoi(argv[3]);
    printf("coding length: %d\n\n", c);
    fwrite("HECC", 4, 1, t);
    sprintf(mc, "%d", c);
    fwrite(mc, 2, 1, t);
    if(c == 8)
    {
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            g = 0;
            for(i = 0; i < 12; i++)
            {
                if((i != 0)&&(i != 1)&&(i != 3)&&(i != 7))
                    kod[i] = byte[g++];
                else
                    kod[i] = 2;
            }
            kontrolbit(kod, 12);
            g = 0;
            while(g < 12)
            {
                for(i = 0; i < 8; i++)
                {
                    if((i == 0)||(i == 1))
                        byte[i] = 0;
                    else
                        byte[i] = kod[g++];
                }
                fputc(todec(byte), t);
            }
            x = fgetc(f);
        }
    }
    else if(c == 12)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        flag = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            if(flag == 0)
            {
                for(i = 0; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                x = fgetc(f);
                tobit(x, byte);
                for(i = 0; i < 4; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                fseek(f, -1, SEEK_CUR);
                flag = 1;
            }
            else if(flag == 1)
            {
                for(i = 4; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                x = fgetc(f);
                tobit(x, byte);
                for(i = 0; i < 8; i++)
                {
                    if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                        kod[g++] = byte[i];
                    else
                    {
                        kod[g] = 2;
                        g++;
                        i--;
                    }
                }
                flag = 0;
            }
            if(g == 17)
            {
                kontrolbit(kod, 17);
                g = 0;
                while(g < 17)
                {
                    for(i = 0; i < 8; i++)
                    {
                        if(i != 7)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
            x = fgetc(f);
        }
    }
    else if(c == 16)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 21)
            {
                kontrolbit(kod, 21);
                g = 0;
                while(g < 21)
                {
                    for(i = 0; i < 8; i++)
                    {
                        if(i == 0)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
            x = fgetc(f);
        }
    }
    else if(c == 32)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 38)
            {
                kontrolbit(kod, 38);
                g = 0;
                while(g < 38)
                {
                    for(i = 0; i < 8; i++)
                    {
                        if((i != 6)&&(i != 7))
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
            x = fgetc(f);
        }
    }
    else if(c == 64)
    {
        fprintf(stderr, "WARNING: with coding length %d end of file is not always processed correctly\n\n", c);
        g = 0;
        x = fgetc(f);
        while(x != EOF)
        {
            tobit(x, byte);
            for(i = 0; i < 8; i++)
            {
                if((g != 0)&&(g != 1)&&(g != 3)&&(g != 7)&&(g != 15)&&(g != 31)&&(g != 63))
                    kod[g++] = byte[i];
                else
                {
                    kod[g] = 2;
                    g++;
                    i--;
                }
            }
            if(g == 71)
            {
                kontrolbit(kod, 71);
                g = 0;
                while(g < 71)
                {
                    for(i = 0; i < 8; i++)
                    {
                        if(i != 7)
                            byte[i] = 0;
                        else
                            byte[i] = kod[g++];
                    }
                    fputc(todec(byte), t);
                }
                g = 0;
            }
            x = fgetc(f);
        }
    }
    else
    {
        fprintf(stderr, "ERROR: bad coding length %d, maybe 8/12/16/32/64\n\n", c);
    }
    fclose(f);
    fclose(t);
    return 0;
}
